cameo/strain_design/heuristic/evolutionary/optimization.py

```diff
@@ -806,3 +845,209 @@ def __init__(self, reaction_db=None, *args, **kwargs):
+      other
+
+   The implementation here differs from that in [1]_ in that we use a general purpose search algorithm rather than
+   formulating the search as a mixed integer linear programming problem.
+
+   References
+   ----------
+   .. [1] King, Zachary A., and Adam M. Feist. “Optimizing Cofactor Specificity of Oxidoreductase Enzymes for the
+      Generation of Microbial Production Strains—OptSwap.” Industrial Biotechnology 9, no. 4 (August 1,
+      2013): 236–46. ---doi:10.1089/ind.2013.0005.
+
+    Parameters
+    ----------
+    model : SolverBasedModel
+       the model to operator on
+    cofactor_id_swaps : list
+       a list of length 2 where the first element contains a list of metabolite identifiers to swap-out by replacing
+       them with the metabolites listed in the second element of the same list, see e.g. `NADH_NADPH` which is also
+       the default
+    candidate_reactions : list
+       reactions to consider for co-factor swap - if not given then search for all reactions that include the given
+       cofactors
+    skip_reactions : list
+       reactions to not consider for co-factor swap, defaults to the objective function if not provided
+    args, kwargs : keyword arguments
+       passed on to super-classes, see in particular `objective_function`, `heuristic_method`, `termination`,
+       `simulation_method`, `wt_reference`, of `HeuristicOptimization` and `max_size` of `HeuristicOptimization.run`
+
+    Examples
+    --------
+    >>> from cameo import models
+    >>> from cameo.strain_design.heuristic.evolutionary.objective_functions import product_yield
+    >>> model = models.bigg.iJO1366
+    >>> model.objective = model.reactions.EX_thr__L_e
+    >>> model.reactions.BIOMASS_Ec_iJO1366_core_53p95M.lower_bound = 0.1
+    >>> py = product_yield(model.reactions.EX_thr__L_e, model.reactions.EX_glc__D_e)
+    >>> swap_optimization = CofactorSwapOptimization(model=model, objective_function=py)
+    >>> swap_optimization.run(max_evaluations=2000, max_size=2)
+    """
+
+    def __init__(self, model, cofactor_id_swaps=NADH_NADPH, candidate_reactions=None,
+                 skip_reactions=None, *args, **kwargs):
+        super(self.__class__, self).__init__(model=model, knockout_evaluator=SwapEvaluator, *args, **kwargs)
+        self.model = SwapperModel(model, cofactor_id_swaps, candidate_reactions, skip_reactions)
+        self._ko_type = SWAP_KNOCKOUT_TYPE
+        self.representation = list(self.model.swapped_reactions.keys())
+        self._decoder = decoders.ReactionKnockoutDecoder(self.representation, self.model)
+
+
+class SwapperModel(SolverBasedModel):
+    """An extension of the `SolverBasedModel` to add functions to easily add reactions that have co-factors swapped
+    etc.
+
+    Attributes
+    ----------
+    model : SolverBasedModel
+       the model to convert
+    cofactor_id_swaps : list
```

Comment:
true, changing to tuple, however since the co-factor pair is interchangeable we don't think of swap_in and swap_out just two pairs that can be changed for each other hence I stick to the unnamed tuple